### ISSUE TITLE
Unique aggregation cartodb_id across tiles

### ISSUE THREAD

**Issue by jgoizueta** [CartoDB/Windshaft-cartodb#923]
## TL;DR;

We should change the aggregation queries to produce unique `cartodb_id` values across tiles (some do not, namely `centroid` and `point-grid`. This is a problem that affects CartoVL.

Doing so will allow us to add an endpoint in the Maps API for obtaining which features are aggregated into an aggregated cartodb_id.

## Details

For the **default** and **point-sample** aggregation placement modes, the `cartodb_id` of each aggregated cluster of features is the minimum cartodb_id value of the features in the cluster.

For **centroind** and **point-grid** placements, it is an arbitrary unique number in each tile. **It repeats accross tiles for different clusters**.

This can be a problem for the client to identify uniquely features across tiles/zoom levels and it also rules out the possibility of implementing an API endpoint to obtain the list of features aggregated in a given cluster.

We should return globally (dataset-wise) unique cluster ids that can be used to refer back to the clustered features (given the zoom level and resolution).

## Alternative solutions

Possible ways of doing this:
* Use `MIN(cartodb_id)` always. This shouldn't add a significant performance overhead. Getting the original cluster from the aggregated ids (plus zoom level and resolution) might not be efficient, though.
* Use a reversible encoding of the cluster grid cell (we could just use the minimum coordinates, and we could add the cell size and avoid requiring passing zoom level and resolution to the API).

If we choose to implement the latter, and go for encoding the cell (coordinates and resolution), so that each aggregated `cartodb_id` identifies uniquely the aggregation cell globally (for any tile/zoom level), the we have this options:
* Use a PostgresSQL text type, or a decimal or bit string (so we don't have a hard limit on the number of bits).
  + We could reach level 32 aggregation (9.3x9,3 mm) using 64 bits plus 5 to keep the grid size (level).
  + This (using non-integer cartod_ids) is probably incompatible with some of our stack (cartojs? builder?)
* Use PostgreSQL BigInts, and only non-negative numbers:
  + We can get up to level 29 (7.5 cm) (29 bits per coordinate + 5 for the resolution = 63 bits)
* Use PostreSQL BigInts with non-negative values that can preserved into a floating point double (JS numeric type) (53-bit significand + sign):
  + Maximumum would be level 24 (2.4 m) (24*2 + 5 = 53 bits)

If we don't encode the resolution in the cartodb_id (so that cartodb_ids are relative to a given zoom level and resolution), we could reach these limits:

* Use PostgreSQL BigInts, and only non-negative numbers:
  + level 31 (1.9 cm) (63 bits)
* Use PostgreSQL BigInts, and allowing for negative numbers:
  + level 32 (9.3 mm) (64 bits)
* Use JS-safe nonnegative values:
  + leve 26 (60 cm) (53 bits)
* Use JS-safe signed values:
  + leve 27 (30 cm) (54 bits)

### Notes

I like the idea of encoding the aggregation cell in a JS-safe, non-negative bigint, but is 2.4x2.4 meters a reasonable aggregation limit?

If we use integer-encoding with higher precision on the client (JS) we'd need to handle these ids as strings because of the floating point precision limits.

A nice feature of using the integer cell encoding is that we can interleave the coordinate bits and have the benefits of quadkeys (e.g. ordering by it preserves spatial locality); if we prefix with the resolution this would work only inside each zoom level, or we cooud use a resolution suffix

Currently we limit the grid size to a minimum of 0.15 mm (level 38), but that value is completely arbitrary, just t avoid divisions by 0.

Regarding adding an API to obtain which features are aggregated into a cluster using the aggregated cartodb_id, since the request would need to refer to a map and layer anyway, the fact of needing the zoom and resolution or not besides the cartodb_id would be irrelevant.

The simple solution of using MIN(cartodb_id) would problably complicate the implementation of such API.

**Comment by jgoizueta** [comment]
🤔 Obtaining the aggregated features given the id of one of them (`MIN(cartodb_id)`) doesn't seem as  hard as I had Imagined:

```javascript
const clusterFeaturesQuery = ctx => `
    WITH
    _cdb_params AS (
        SELECT
        ${gridResolution(ctx)} AS res
    ),
    _cell AS (
        SELECT
        ST_MakeEnvelope(
            Floor(ST_X(_cdb_query.the_geom_webmercator)/_cdb_params.res)*_cdb_params.res,
            Floor(ST_Y(_cdb_query.the_geom_webmercator)/_cdb_params.res)*_cdb_params.res,
            Floor(ST_X(_cdb_query.the_geom_webmercator)/_cdb_params.res + 1)*_cdb_params.res,
            Floor(ST_Y(_cdb_query.the_geom_webmercator)/_cdb_params.res + 1)*_cdb_params.res,
            3857
        ) AS bbox
        FROM (${ctx.sourceQuery}) _cdb_query
        WHERE _cdb_query.cartodb_id = ${ctx.id}
    )
    SELECT _cdb_query.* FROM _cell, (${ctx.sourceQuery}) _cdb_query
        WHERE ST_Intersects(_cdb_query.the_geom_webmercator, _cell.bbox
`;
```

This query should benefit from the spatial index on `the_geom_webmercator` (and the PK index on `cartodb_id`).

I'll check the performance and validity of this query and if it seems OK I think we might just implement the easy solution (MIN(cartodb_id)) since it will allow querying the individual features efficiently.

**Comment by jgoizueta** [comment]
This has been fixed for a long time
